### PR TITLE
[UNR-5430] Gracefully shutdown runtime on windows via process signal

### PR DESCRIPTION
**Pull request by MCArth-improbable**
#### Description
Previously we would request for the runtime to shut down via a net message. That was quite slow. This PR adds the ability to request the graceful shutdown via a process signal, which is slightly less slow.

#### Release note
`- The runtime will shut down slightly faster after a PIE session.`

#### Tests
I tested that the runtime still gracefully shut down as expected - and tested that it shut down slightly faster (-1s avg from this PR, n=2)

What automated tests are included in this PR?
none (apart from existing editor tests!)

STRONGLY SUGGESTED: How can this be verified by QA?
by starting and stopping play in editor, along with closing the editor while still playing

#### Documentation
release note

#### Primary reviewers
@MatthewSandfordImprobable

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 1]
```diff
@@ -27,7 +21,7 @@
 #include "SpatialCommandUtils.h"
 #include "SpatialGDKServicesConstants.h"
 #include "SpatialGDKServicesModule.h"
-#include "UObject/CoreNet.h"
+#include "Windows/MinWindows.h"
```
Comment (MCArth-improbable) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: So I'm a bit worried about this. Including this header leads to `winnt.h(603): [C4005] 'TEXT': macro redefinition` compilation warning which seems like it might break quite a bit of stuff?

Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: is this actually a problem?

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 2]
```diff
@@ -2,22 +2,16 @@
 
 #include "LocalDeploymentManager.h"
 
-#include "AssetRegistryModule.h"
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:5]: Are these removed as they are already included via other includes or are these removed as we do not use any code from these headers?

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 1 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: Mmmm, I think we ran into this before. We should find a solution to this but... I am not sure what that is...

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 2 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:5]: don't use any code from those headers

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 1 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: seems like all the stuff we need is in "WinUser.h" but when I only import that I get the error `  LocalDeploymentManager.cpp(506): [C3861] 'SendMessage': identifier not found` so I assume it has dependencies on some of the other files included in MinWindows :/

File: SpatialGDK/Source/SpatialGDKServices/Public/LocalDeploymentManager.h [hunk 3]
```diff
@@ -68,6 +70,8 @@ class FLocalDeploymentManager final
 
 	bool WaitForRuntimeProcessToShutDown();
 	bool StartLocalDeploymentShutDown();
+	bool WaitForGracefulShutdown();
+	bool WaitForForcefulShutdown();
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Public/LocalDeploymentManager.h:74]: I might prefer something like ForceRuntimeShutdown. To me WaitForForcefulShutdown does not suggest that it actual does the shutting down. What do you think?

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 1 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: This header won't exist on non windows platforms

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: Is there no platform agnostic facilities that unreal offers that we can use? If not I suggest sticking this in a platform layer of our own, I'd prefer it if we silo'd platform specific code if possible.

File: SpatialGDK/Source/SpatialGDKServices/Public/LocalDeploymentManager.h (hunk 3 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Public/LocalDeploymentManager.h:75]: hmm, good point. My intention was to communicate that it'd 'lock' the thread in a spin lock, waiting for the process to die. I feel like that's maybe worthwhile information to get across at the level of abstraction it's used. Maybe something like `ForceShutdownThenWaitForCompletion` would work?

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 4]
```diff
@@ -500,6 +489,53 @@ void FLocalDeploymentManager::FinishLocalDeploymentShutDown()
...
+	{
+		UE_LOG(LogSpatialDeploymentManager, Error,
+			   TEXT("Tried to gracefully stop local deployment but could not find runtime window. Runtime will eventually stop "
+					"non-gracefully."));
+	}
+}
+
+void FLocalDeploymentManager::NonWindowsRequestGracefulRuntimeShutdown()
+{
+	// Sends a http message requesting graceful shutdown
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:516]: Why don't we do any of this stuff on windows anymore?

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:516]: Also we don't try close any windows on non windows platforms? Do we just not care?

Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:516]: process shutdown is slightly faster, but implementation is different on mac (i.e. nonwindows). Will leave a comment in the code on this

Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:516]: r.e. second question, the essence of what we're doing here is gracefully shutting down the runtime. not actually sure why we're doing that via a 'window' I got the code from @MatthewSandfordImprobable (he previously had a look at doing this but other things then meant it didn't work). Would be interested why we're actually doing this matt

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 5]
```diff
@@ -473,6 +444,24 @@ bool FLocalDeploymentManager::StartLocalDeploymentShutDown()
...
+	NonWindowsRequestGracefulRuntimeShutdown();
+#endif
+	bool bGracefulShutdownSuccess = WaitForRuntimeProcessToShutDown();
+	return bGracefulShutdownSuccess;
+}
+
+bool FLocalDeploymentManager::WaitForForcefulShutdown()
+{
+	RuntimeProcess->Stop();
+	const bool bForcefulShutdownSuccess = WaitForRuntimeProcessToShutDown();
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:461]: Nit: Think we can remove this line and just return WaitForRuntimeProcessToShutDown.
I think the reasons to have is as a separate line would be either to cache the value, readability or debuggability. I don't think those are aided in this case.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 6]
```diff
@@ -473,6 +444,24 @@ bool FLocalDeploymentManager::StartLocalDeploymentShutDown()
 	return true;
 }
 
+bool FLocalDeploymentManager::WaitForGracefulShutdown()
+{
+#if PLATFORM_WINDOWS
+	WindowsRequestGracefulRuntimeShutdown();
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:450]: Nit: Would prefer if these were explicit on how they shut the runtime down. For example ShutdownRuntimeWithProcessSignal / ShutdownRuntimeWithHTTPRequest. What do you think?

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 7]
```diff
@@ -402,12 +396,10 @@ bool FLocalDeploymentManager::TryStopLocalDeployment()
...
-
-	bool bRuntimeShutDownSuccesfully = WaitForRuntimeProcessToShutDown();
+	const bool bRuntimeShutDownSuccessfully = WaitForForcefulShutdown();
 	FinishLocalDeploymentShutDown();
 
-	return bRuntimeShutDownSuccesfully;
+	return bRuntimeShutDownSuccessfully;
 }
 
 bool FLocalDeploymentManager::TryStopLocalDeploymentGracefully()
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:405]: Like the new structure of this function. Much clearer. NICE!

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 8]
```diff
@@ -500,6 +489,53 @@ void FLocalDeploymentManager::FinishLocalDeploymentShutDown()
 	bStoppingDeployment = false;
 }
 
+void FLocalDeploymentManager::WindowsRequestGracefulRuntimeShutdown()
+{
+	// Sends a process signal requesting the runtime shutdown
+
+	// Prepare runtime process name
+	FString RuntimeProcessName = RuntimePath;
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:497]: Should be able to do:
FString RuntimeProcessName = RuntimePath.Replace(TEXT("/"), TEXT("\\"));

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 4 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:516]: oh, having read a bit more about windows signals it seems like you need a window handle to call WM_CLOSE (which is nicer than terminateprocess)

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 1 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: Ideally we would use some unreal library to allow us to work cross platform. I believe last time this was investigated we did find some promising unreal code but no working solution.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 2 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:5]: Cool, if that is the case we can remove these. We usually try not to rely on headers being included from other headers and ensure we are explicit about every header we need.

File: SpatialGDK/Source/SpatialGDKServices/Public/LocalDeploymentManager.h (hunk 3 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Public/LocalDeploymentManager.h:75]: Yeah that sounds reasonable.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 4 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:516]: If we have a way to close the process on mac super! I had not luck when I was searching. I tried some non platform specific methods to close the process but they all seemed to terminate the process rather than send a close signal.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 1 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: How about: `void SpatialCommandUtils::TryKillProcessWithName(const FString& ProcessName)`

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 8 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:497]: could even make it const

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 1 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: I believe that terminates the process rather than sending a close signal? I could be wrong though...?

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: hat tip Dmitrii

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: Thinking about it a bit more maybe what you want is `FGenericWindow::Destroy()`

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: How about leveraging: `FPlatformProcess::CloseProc`

Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: seems like closeproc isn't what we want either, looks like it was discussed in matt's pr https://github.com/spatialos/UnrealGDK/pull/2899

Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: though that might have changed as i think the reason it ended up being a net message was because there wasn't process signal support. It doesn't really sound like it requests for the program to close though, from the docs https://docs.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle
@MatthewSandfordImprobable any thoughts, reckon this might work?

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 4 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:516]: @MatthewSandfordImprobable I should've asked actually, is the runtime listening for a SIGTERM signal or something on unix? I don't really know how WM_CLOSE maps to unix

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 1 above)
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: > Some success here but after more testing I am a little worried CloseProc isn't doing what we want it to. So calling SpatialCommandUtils::TryKillProcessWithName after CloseProc will actually terminate it. Brilliant. I am still not super convinced that the waiting logic is now valid. Also still concerned CloseProc does not actually do what we want it to do.

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: might be worth considering.

Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:24]: We might be able to look at the runtime logs and see if it prints out that it received an exit signal? This might help working out if we are closing or terminating.

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 9]
```diff
@@ -58,62 +59,6 @@ bool SpatialCommandUtils::AttemptSpatialAuth(bool bIsRunningInChina)
...
-
-	bool bSuccess = OutExitCode == 0;
-	if (!bSuccess)
-	{
-		UE_LOG(LogSpatialCommandUtils, Warning, TEXT("Spatial stop service failed. Error Code: %d, Error Message: %s"), OutExitCode,
-			   *OutResult);
-	}
-
-	return bSuccess;
-}
```
Comment (MCArth-improbable) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:115]: drive-by removal of unused functions

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 10]
```diff
@@ -7,6 +7,7 @@
 #include "Serialization/JsonSerializer.h"
 #include "SpatialGDKServicesConstants.h"
 #include "SpatialGDKServicesModule.h"
+#include "Windows/AllowWindowsPlatformTypes.h"
```
Comment (MCArth-improbable) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:10]: Is this the right place for the header? I know it's alphabetical, but don't know if that's based on path initial letter or file name initial letter

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 8 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp]: changed!

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 6 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp]: Changed to use `TryGracefullyKill` and `TryGracefullyKillWindows` within spatialcommandutils in updated PR, I think the windows name makes more sense with this update as now both methods send a signal?

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 5 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:463]: Hmm, I included this as I felt it wasn't super clear what WaitForRuntimeProcessToShutDown was returning. do you think the tradeoff of a slightly less clear function is outweighed by a shorter function?

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 1 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp]: I did some more digging (with miron)

So it seems like `TerminateProc` on windows isn't what we want, as on windows it calls the windows API `TerminateProcess`, which immediately kills the process. However, on unix, TerminateProc sends a SIGTERM signal which I believe is what we want (but don't really have a way to test at the moment). So I've refactored the methods to SpatialCommandUtils.h and are using WM_CLOSE on windows, while using terminateproc on unix

lmk what you think

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp (hunk 10 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:10]: clang-format will place it in the correct place, I believe

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp (hunk 5 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:463]: On second read, I think you are right and it does help readability!

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp (hunk 9 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:115]: POG!

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 11]
```diff
@@ -449,6 +378,34 @@ void SpatialCommandUtils::TryKillProcessWithName(const FString& ProcessName)
 	}
 }
 
+void SpatialCommandUtils::TryGracefullyKill(FString ProcName, FProcHandle ProcHandle)
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:381]: FString - > const FString&

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 12]
```diff
@@ -449,6 +378,34 @@ void SpatialCommandUtils::TryKillProcessWithName(const FString& ProcessName)
...
+	const HWND RuntimeWindowHandle = FindWindowA(nullptr, TCHAR_TO_ANSI(*ProcName));
+	if (RuntimeWindowHandle != nullptr)
+	{
+		// Using SendMessageW to send message despite being 'supposed to' use SendMessage - as including "Windows/MinWindows.h" needed for
+		// SendMessage overrides the TEXT macro (SendMessageW is also used elsewhere in unreal)
+		SendMessageW(RuntimeWindowHandle, WM_CLOSE, NULL, NULL);
+	}
+	else
+	{
+		UE_LOG(LogSpatialDeploymentManager, Error, TEXT("Tried to gracefully stop process but could not find runtime window."));
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:405]: Might be cute to include the process name in the log.

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp (hunk 11 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:381]: FProcHandle  could also be a const ref

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp (hunk 12 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp]: good point! added

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp (hunk 11 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp]: changed!

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 13]
```diff
@@ -388,9 +388,9 @@ void SpatialCommandUtils::TryGracefullyKill(FString ProcName, FProcHandle ProcHa
 #endif
 }
 
-void SpatialCommandUtils::TryGracefullyKillWindows(FString ProcName)
+void SpatialCommandUtils::TryGracefullyKillWindows(const FString ProcName)
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:391]: Should also be a ref

Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp]: added!

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 14]
```diff
@@ -449,6 +384,38 @@ void SpatialCommandUtils::TryKillProcessWithName(const FString& ProcessName)
...
+#else
+	// TerminateProc is too forceful on Windows
+	TryGracefullyKillWindows(ProcName);
+#endif
+}
+
+void SpatialCommandUtils::TryGracefullyKillWindows(const FString& ProcName)
+{
+#if PLATFORM_WINDOWS
+	// Use WM_CLOSE signal on windows as DestroyProc forcefully kills on Windows
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:401]:
```suggestion
	// Use WM_CLOSE signal on windows as TerminateProc forcefully kills on Windows
```
?

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 15]
```diff
@@ -449,6 +384,38 @@ void SpatialCommandUtils::TryKillProcessWithName(const FString& ProcessName)
...
+
+	// Find runtime window
+	const HWND RuntimeWindowHandle = FindWindowA(nullptr, TCHAR_TO_ANSI(*ProcName));
+	if (RuntimeWindowHandle != nullptr)
+	{
+		// Using SendMessageW to send message despite being 'supposed to' use SendMessage - as including "Windows/MinWindows.h" needed for
+		// SendMessage overrides the TEXT macro (SendMessageW is also used elsewhere in unreal)
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:408]:
```suggestion
		// SendMessage overrides the TEXT macro (SendMessageW is also used elsewhere in Unreal)
```
very important

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 16]
```diff
@@ -8,6 +8,13 @@
 #include "SpatialGDKServicesConstants.h"
 #include "SpatialGDKServicesModule.h"
 
+#if PLATFORM_WINDOWS
+#include "Windows/AllowWindowsPlatformTypes.h"
+#include "Windows/HideWindowsPlatformTypes.h"
+#endif
+
+#include <csignal>
+
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:17]:
```suggestion
```

Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:16]: whoops

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp (hunk 14 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:400]: ?

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 17]
```diff
@@ -449,6 +380,38 @@ void SpatialCommandUtils::TryKillProcessWithName(const FString& ProcessName)
...
+#if !PLATFORM_WINDOWS
+	// On UNIX this sends a SIGTERM signal
+	FGenericPlatformProcess::TerminateProc(ProcHandle);
+#else
+	// TerminateProc is too forceful on Windows
+	TryGracefullyKillWindows(ProcName);
+#endif
+}
+
+void SpatialCommandUtils::TryGracefullyKillWindows(const FString& ProcName)
```
Comment (MatthewSandfordImprobable) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:394]: Might be preferable to have the whole function in the #if PLATFORM_WINDWOS

You will need to add the check in the header as well.

**Review by MatthewSandfordImprobable — APPROVED** [review]

**Review by simonsarginson — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp [hunk 18]
```diff
@@ -449,6 +380,38 @@ void SpatialCommandUtils::TryKillProcessWithName(const FString& ProcessName)
...
 
+void SpatialCommandUtils::TryGracefullyKill(const FString& ProcName, const FProcHandle& ProcHandle)
+{
+#if !PLATFORM_WINDOWS
+	// On UNIX this sends a SIGTERM signal
+	FGenericPlatformProcess::TerminateProc(ProcHandle);
+#else
+	// TerminateProc is too forceful on Windows
+	TryGracefullyKillWindows(ProcName);
+#endif
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:391]: very minor: generally its considered good practice to do positive case first, reads a bit easier.
```suggestion
#if PLATFORM_WINDOWS
	// TerminateProc is too forceful on Windows
	TryGracefullyKillWindows(ProcName);
#else
	// On UNIX this sends a SIGTERM signal
	// TODO: does this work on Mac?
	FGenericPlatformProcess::TerminateProc(ProcHandle);
#endif
```

Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:391]: yeah that is a fair point

Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:391]: Nice catch with the `FGenericPlatformProcess`! I think I agree with you on the positive case being first as well, will commit the suggestion. Thanks!

File: SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp (hunk 17 above)
Comment (MCArth-improbable) [reply SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp:396]: added!

**Review by simonsarginson — APPROVED** [review]
